fplll/pruner.h

```diff
@@ -0,0 +1,568 @@
+
+
+
+template<class FT> 
+class Pruner{
+  public:
+    class TestPruner;
+    friend class TestPruner;
+
+    FT preproc_cost;
+    FT target_success_proba;
+    FT enumeration_radius;
+
+    
+    Pruner();
+
+    template<class GSO_ZT,class GSO_FT>
+    void load_basis_shape(MatGSO<GSO_ZT, GSO_FT>& gso, int beginning = 0, int end = 0);
+    void load_basis_shape(int dim, double* gso_sq_norms);
+    
+    //void set_parameters(FT preproc_cost, FT target_success_proba);
+    
+    void optimize_pruning_coeffs(/*io*/double* pr, /*i*/ int reset = 1);
+    double get_enum_cost(/*i*/double* pr);
+    double get_enum_cost_with_retrials(/*i*/double* pr);
+    double get_svp_success_proba(/*i*/double* pr);
+
+
+
+  private:
+    using vec = array<FT, PRUNER_MAX_N>;
+    using evec = array<FT, PRUNER_MAX_D>; 
+          // Even vectors, i.e. only one every two entry is stored: V[2i] = V[2i+1] =E[i]
+    using poly = array<FT, PRUNER_MAX_D + 1>;
+
+
+    void set_tabulated_consts();
+    int n;  // Dimension of the (sub)-basis
+    int d;  // Degree d = floor(n/2)
+
+    vec r;
+    vec pv;
+    FT renormalization_factor;
+
+
+    int check_loaded_basis();
+    void init_prunning_coeffs(evec &b);
+    void load_prunning_coeffs(/*i*/double* pr, /*o*/ evec& b);
+    void save_prunning_coeffs(/*o*/double* pr, /*i*/ evec& b);
+    inline int enforce(/*io*/ evec &b, /*opt i*/ int j = 0);
+    inline FT eval_poly(int ld,/*i*/ poly *p, FT x);
+    inline void integrate_poly(int ld,/*io*/ poly *p);
+    inline FT relative_volume(/*i*/int rd, evec &b);
+    inline FT cost(/*i*/ evec &b);
+    inline FT svp_success_proba(/*i*/ evec &b);
+    inline FT cost_factor(/*i*/ evec &b);
+    FT cost_factor_derivative(/*i*/ evec &b, /*o*/ evec &res);
+    int improve(/*io*/ evec &b);
+    void descent(/*io*/ evec &b);
+
+    FT tabulated_factorial[PRUNER_MAX_N];
+    FT tabulated_ball_vol[PRUNER_MAX_N];
+    
+    FT one;  // HACK: because we don't have (double - FT) (yet ?)
+    FT minus_one; // HACK: same here
+  
+
+    FT epsilon;    // Epsilon to use for numerical differentiation
+    FT min_step;    // Minimal step in a given direction 
+    FT min_cf_decrease;    // Maximal ratio of two consectuive cf in the descent before stopping
+
+    FT step_factor; // Increment factor for steps in a given direction
+    FT shell_ratio; // Shell thickness Ratio when evaluating svp proba
+    FT symmetry_factor; // Set at 2 for SVP enumeration assuming the implem only explore half the space
+};
+
+template<class FT>
+Pruner<FT>::Pruner(){
+  n = 0;
+  d = 0;
+  set_tabulated_consts();
+  epsilon = std::pow(2., -13);
+  min_step = std::pow(2., -12);
+  step_factor = std::pow(2, .5);
+  shell_ratio = .995;
+  minus_one = -1.;
+  one = 1.;
+  preproc_cost = 0.;
+  enumeration_radius = 0.;
+  target_success_proba = .90;
+  preproc_cost = 0;
+  min_cf_decrease = .9999;
+  symmetry_factor = 2;
+}
+
+
+template<class FT>
+void Pruner<FT>::set_tabulated_consts(){
+  mpfr_t tmp;
```

Comment:
It would be nice to have a read_from_string function in the FP_NR class indeed. Adding it as an issue.